### PR TITLE
feat: recipe image upload

### PR DESCRIPTION
**Pull request by marybrown123**

### CODE REVIEW

File: src/recipe/file.service.ts [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+import { ConfigService } from '@nestjs/config';
+import { S3Client, PutObjectCommand } from '@aws-sdk/client-s3';
+
+@Injectable()
+export class FileService {
+  private s3Client = new S3Client({
+    region: this.configService.getOrThrow('AWS_S3_REGION'),
+    credentials: {
+      accessKeyId: process.env.AWS_ACCESS_KEY_ID,
+      secretAccessKey: process.env.AWS_SECRET_ACCESS_KEY,
```
Comment (wojdanowski) [inline src/recipe/file.service.ts:11]: You could use configService to get this envs too

File: src/recipe/file.service.ts [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+      Body: file,
+    };
+    const command = new PutObjectCommand(params);
+    try {
+      await this.s3Client.send(command);
+    } catch (error) {
+      throw new ForbiddenException();
+    }
+
+    return `https://${params.Bucket}.s3.${process.env.AWS_S3_REGION}.amazonaws.com/${params.Key}`;
```
Comment (wojdanowski) [inline src/recipe/file.service.ts:30]: You could use configService to get this envs too

File: src/recipe/file.service.ts [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+      accessKeyId: process.env.AWS_ACCESS_KEY_ID,
+      secretAccessKey: process.env.AWS_SECRET_ACCESS_KEY,
+    },
+  });
+
+  constructor(private configService: ConfigService) {}
+
+  async uploadFileToS3(fileName: string, file: Buffer) {
+    const params = {
+      Bucket: process.env.S3_BUCKET_NAME,
```
Comment (wojdanowski) [inline src/recipe/file.service.ts:19]: You could use configService to get this envs too

File: src/recipe/recipe.controller.ts [hunk 4]
```diff
@@ -26,23 +28,41 @@ import {
...
 
   @Post()
+  @UseInterceptors(FileInterceptor('file'))
   @UseGuards(AuthGuard('jwt'))
   @ApiOperation({ summary: 'Create a new recipe' })
   @ApiCreatedResponse({ type: RecipeResponse })
   @ApiUnauthorizedResponse({ description: 'Not logged in' })
   async createRecipe(
     @Body() recipe: CreateRecipeDTO,
+    @UploadedFile() file: Express.Multer.File,
```
Comment (wojdanowski) [inline src/recipe/recipe.controller.ts:45]: Why do you need this on createRecipe when you have a different endpoint to upload an image?

File: src/recipe/commands/handlers/uploadRecipeImage.handler.ts [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+    await this.recipeDAO.findRecipeById(recipeId);
+
+    const imageURL = await this.fileService.uploadFileToS3(
+      file.originalname,
+      file.buffer,
+    );
+
+    const recipeWithImage: UpdateRecipeDTO = { imageURL };
+
+    return this.recipeDAO.updateRecipe(recipeId, recipeWithImage);
```
Comment (wojdanowski) [inline src/recipe/commands/handlers/uploadRecipeImage.handler.ts:24]: have you tried to fetch this recipe to see if the image is available? (having this imageUrl that you save on recipe)

File: src/recipe/file.service.ts [hunk 6]
```diff
@@ -0,0 +1,32 @@
+import { ForbiddenException, Injectable } from '@nestjs/common';
+import { ConfigService } from '@nestjs/config';
+import { S3Client, PutObjectCommand } from '@aws-sdk/client-s3';
+
+@Injectable()
+export class FileService {
```
Comment (wojdanowski) [inline src/recipe/file.service.ts:6]: I would go with a separate service for S3 service that would establish configuration (secrets, keys etc) and expose a method to upload the file (directly using `S3Client`)

The other service would be FileService that would be responsible for business logic of file processing. And would call the previous service to upload the files.

**Important note. Service that handles s3 configuration should be mocked!!!**

**Review by wojdanowski — CHANGES_REQUESTED** [review]
General notes:

- You have added logic that is not covered with tests!
- Please verify if you are able to upload the file and open the uploaded image with provided URL

File: src/recipe/recipe.controller.ts [hunk 7]
```diff
@@ -26,23 +28,41 @@ import {
...
     @CurrentUser() user: User,
   ): Promise<RecipeResponse> {
     return this.recipeService.createRecipe(recipe, user.id);
   }
 
+  @Post('/upload/:id')
+  @UseInterceptors(FileInterceptor('file'))
```
Comment (wojdanowski) [inline src/recipe/recipe.controller.ts:52]: Have you tested this endpoint manually with postman?
Hit use:
multipart/form-data as body

Comment (marybrown123) [reply src/recipe/recipe.controller.ts:52]: yes i did

File: src/recipe/file.service.ts (hunk 1 above)
Comment (marybrown123) [reply src/recipe/file.service.ts]: Done

File: src/recipe/file.service.ts (hunk 2 above)
Comment (marybrown123) [reply src/recipe/file.service.ts]: Done

File: src/recipe/file.service.ts (hunk 3 above)
Comment (marybrown123) [reply src/recipe/file.service.ts]: Done

File: src/recipe/recipe.controller.ts (hunk 4 above)
Comment (marybrown123) [reply src/recipe/recipe.controller.ts:45]: i don't need it, my mistake, i removed this line

File: src/recipe/commands/handlers/uploadRecipeImage.handler.ts (hunk 5 above)
Comment (marybrown123) [reply src/recipe/commands/handlers/uploadRecipeImage.handler.ts:23]: What exactly do you mean by this?

File: src/recipe/file.service.ts (hunk 6 above)
Comment (marybrown123) [reply src/recipe/file.service.ts:6]: I think it is done

File: src/recipe/commands/handlers/uploadRecipeImage.handler.ts (hunk 5 above)
Comment (wojdanowski) [reply src/recipe/commands/handlers/uploadRecipeImage.handler.ts:23]: Are the files that you upload to s3 public? Are you able to download them as a client using imageURL?

File: src/recipe/recipe.controller.ts (hunk 4 above)
Comment (wojdanowski) [reply src/recipe/recipe.controller.ts:45]: Still not removed

File: src/recipe/file.service.ts [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+  constructor(
+    private configService: ConfigService,
+    private s3Service: S3Service,
+  ) {}
+
+  async uploadFileToS3(file: Express.Multer.File) {
+    const bucket = this.configService.getOrThrow('S3_BUCKET_NAME');
```
Comment (wojdanowski) [inline src/recipe/file.service.ts:13]: I think the better approach would be use bucket name from envs and add fallback value if it's to present. Egample:

```suggestion
    const bucket = this.configService.get('S3_BUCKET_NAME') || 'default_bucket_name';
```

File: src/recipe/test/unit/file.service.spec.ts [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+      imports: [AppModule],
+    }).compile();
+
+    await module.createNestApplication().init();
+    s3Service = module.get<S3Service>(S3Service);
+    fileService = module.get<FileService>(FileService);
+  });
+
+  it('should upload image to s3', async () => {
+    const s3ServiceUpload = jest
```
Comment (wojdanowski) [inline src/recipe/test/unit/file.service.spec.ts:37]: You should add s3Service mock and use it in the tests

File: src/recipe/file.service.ts (hunk 8 above)
Comment (marybrown123) [reply src/recipe/file.service.ts:13]: What is default_bucket_name

Comment (wojdanowski) [reply src/recipe/file.service.ts:13]: just an example. The idea is that you use bucket name from env variables. If this is not present (this particular env variable was not set in the system) then you hardcode a default value. 
In your case, it should be your bucket name

**Review by wojdanowski — APPROVED** [review]

**Pull request merged**